backend/scripts/test_citation_builder.py

```diff
@@ -0,0 +1,281 @@
+    log = False
+
+    def __init__(
+        self,
+        test_name: str,
+        input_text: str,
+        expected_text: str,
+        expected_citations: List[str],
+    ):
+        self.test_name = test_name
+        self.input_text = input_text
+        self.expected_text = expected_text
+        self.expected_citations = [
+            CitationInfo(citation_num=mock_doc_mapping[doc_id], document_id=doc_id)
+            for doc_id in expected_citations
+        ]
+
+    def run_test(self, mock_data):
+        from test_citation_builder import process_text  # Import the existing function
+
+        final_answer_text, citations = process_text(self.input_text, mock_data)
+        if SimpleCitationExtractionTestCase.log:
+            print(f"Test: {self.test_name}")
+            print(f"Input: {self.input_text}")
+            print(f"Expected: {self.expected_text}")
+            print(f"Got: {final_answer_text}")
+
+        assert final_answer_text.strip() == self.expected_text.strip(), (
+            f"Test '{self.test_name}' failed: Final answer text does not match expected output.\n"
+            f"Expected: {self.expected_text}\nGot: {final_answer_text}"
+        )
+
+        assert citations == self.expected_citations, (
+            f"Test '{self.test_name}' failed: Citations do not match expected output.\n"
+            f"Expected: {self.expected_citations}\nGot: {citations}"
+        )
+
+        if SimpleCitationExtractionTestCase.log:
+            print(f"Test '{self.test_name}' passed successfully!")
+
+        print(f"Test '{self.test_name}' passed.")
+        return True
+
+import unittest
+import random
+from datetime import datetime
+from typing import List
+
+
+
+
+class TestCitationExtraction(unittest.TestCase):
+    @classmethod
+    def setUpClass(cls):
+        cls.mock_data = (mock_docs, mock_doc_mapping)
+
+    def run_test_case(self, test_case):
+        with self.subTest(test_case.test_name):
+            self.assertTrue(test_case.run_test(self.mock_data))
+
+    def test_single_citation(self):
+        test_case = SimpleCitationExtractionTestCase(
+            test_name="Single citation",
+            input_text="""Growth! [1].""",
+            expected_text="""Growth! [[0]](https://0.com).""",
+            expected_citations=["doc_0"],
+        )
+        self.run_test_case(test_case)
+
+
+    def test_multiple_unique_citations(self):
+        test_case = SimpleCitationExtractionTestCase(
+            test_name="Multiple unique citations",
+            input_text=""""Test! [1] [5].""",
+            expected_text=""""Test! [[0]](https://0.com) [[2]](https://2.com).""",
+            expected_citations=["doc_0", "doc_2"],
+        )
+        self.run_test_case(test_case)
+
+    def test_repeated_citations(self):
+        test_case = SimpleCitationExtractionTestCase(
+            test_name="Repeated citations",
+            input_text="""Test! [1][1][1]. And some more [1][2].""",
```

Comment:
across the board, we should test by specifying tokens explicitly by passing in list of strings (tokens) rather than having the tokenizing function